### ISSUE TITLE
ArgumentOutOfRangeException in 2.0.0 GA version

### ISSUE THREAD

**Issue by cdmihai** [PowerShell/PSReadLine#1469]
Error message said I should report this so here it goes:

```
E:\projects\MSBuildStaticGraphTester [transitiveReferences +7 ~5 -0 !]> .\src\out\Release\bin\GraphGen\net472\GraphGen.exe E:\projects\roslyn\src\NuGet\Microsoft.Net.Compilers.Toolset\Microsoft.Net.Compilers.Toolset.Package.csproj package.pdf "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\Mstart .\package_PathsEndingIn_vbc.csproj.pdf
Oops, something went wrong.  Please report this bug with the details below.
Report on GitHub: https://github.com/PowerShell/PSReadLine/issues/new
### Environment
PSReadLine: 2.0.0
PowerShell: 7.0.0
OS: Microsoft Windows 10.0.18363
Last 200 Keys
```
```
 Enter
 s t a r t Spacebar p a Tab Tab Enter
 UpArrow UpArrow DownArrow DownArrow Spacebar Backspace Backspace Backspace g i t Spacebar s t a t u s Enter
 . . \ d e Tab Spacebar Tab Tab Backspace Backspace Backspace Backspace Backspace Backspace Backspace Backspace Backspace Backspace Backspace Backspace Backspace Backspace Backspace Backspace - Tab Spacebar E n t e r p r i s e Enter
 UpArrow UpArrow UpArrow UpArrow Enter
 s t a r t Spacebar p a Tab Tab Enter
 UpArrow UpArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+RightArrow Ctrl+RightArrow Ctrl+RightArrow " " Spacebar LeftArro
w LeftArrow Ctrl+v Enter
 UpArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+RightArrow Ctrl+RightArrow Ctrl+RightArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow
 Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+RightArrow Ctrl+LeftArrow Ctrl+RightArrow Ctrl+RightArrow Ctrl+RightArrow Ctrl+RightArrow Ctrl+RightArrow Ctrl
+RightArrow Ctrl+RightArrow Ctrl+RightArrow Ctrl+RightArrow Ctrl+RightArrow Ctrl+RightArrow Ctrl+RightArrow Ctrl+RightArrow Ctrl+RightArrow Ctrl+RightArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow RightArrow RightArro
w RightArrow RightArrow LeftArrow Shift+Ctrl+LeftArrow Shift+Ctrl+LeftArrow Ctrl+x Backspace Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+Lef
tArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+RightArrow LeftArrow Ctrl+v Spacebar Enter
 s t a r t Spacebar p a c k Tab Tab Enter
 UpArrow UpArrow DownArrow DownArrow UpArrow UpArrow DownArrow
```

### Exception
```
System.ArgumentOutOfRangeException: The value must be greater than or equal to zero and less than the console's buffer size in that dimension. (Parameter 'top')
Actual value was 55.
   at System.ConsolePal.SetCursorPosition(Int32 left, Int32 top)
   at Microsoft.PowerShell.PSConsoleReadLine.ReallyRender(RenderData renderData, String defaultColor)
   at Microsoft.PowerShell.PSConsoleReadLine.ForceRender()
   at Microsoft.PowerShell.PSConsoleReadLine.UpdateFromHistory(HistoryMoveCursor moveCursor)
   at Microsoft.PowerShell.PSConsoleReadLine.HistoryRecall(Int32 direction)
   at Microsoft.PowerShell.PSConsoleReadLine.NextHistory(Nullable`1 key, Object arg)
   at Microsoft.PowerShell.PSConsoleReadLine.ProcessOneKey(PSKeyInfo key, Dictionary`2 dispatchTable, Boolean ignoreIfNoAction, Object arg)
   at Microsoft.PowerShell.PSConsoleReadLine.InputLoop()
   at Microsoft.PowerShell.PSConsoleReadLine.ReadLine(Runspace runspace, EngineIntrinsics engineIntrinsics, CancellationToken cancellationToken)
```

I was going through the command history with the up / down keys when this happened.

**Comment by iSazonov** [comment]
See #1306

**Comment by petacreepers23** [comment]
I had same error today, but with parameter "top". also reported as it said

**Comment by daxian-dbw** [comment]
This happened with PSReadLine 2.0.0, so it's something new.

@cdmihai Can you please run the script below in the pwsh session where you run into this issue to collect the environment data?
Also, are you able to consistently reproduce the issue? If so, can you please provide the repro steps (including your history commands that triggers the issue when using `up/down` keys)?

```powershell
& {
    $hostName = $Host.Name
    if ($hostName -eq "ConsoleHost" -and (Get-Command Get-CimInstance -ErrorAction SilentlyContinue)) {
        $id = $PID
        $inWindowsTerminal = $false
        while ($true) {
            $p = Get-CimInstance -ClassName Win32_Process -Filter "ProcessId Like $id"
            if (!$p -or !$p.Name) { break }
            if ($p.Name -eq "WindowsTerminal.exe") { $inWindowsTerminal = $true; break }
            $id = $p.ParentProcessId
        }
        if ($inWindowsTerminal) { $hostName += " (Windows Terminal)" }
    }

    "`nPS version: $($PSVersionTable.PSVersion)"
    $m = Get-Module PSReadline
    $v = $m.Version; $pre = $m.PrivateData.PSData.Prerelease
    if ($pre) { $v = "$v-$pre" }
    "PSReadline version: $v"
    if ($IsLinux -or $IsMacOS) {
        "os: $(uname -a)"
    } else {
        "os: $((dir $env:SystemRoot\System32\cmd.exe).VersionInfo.FileVersion)"
    }
    "PS file version: $($name = if ($PSVersionTable.PSEdition -eq "Core") { "pwsh.dll" } else { "powershell.exe" }; (dir $pshome\$name).VersionInfo.FileVersion)"
    "HostName: $hostName"
    "BufferWidth: $([console]::BufferWidth)"
    "BufferHeight: $([console]::BufferHeight)`n"
}
```

**Comment by iSazonov** [comment]
@daxian-dbw Maybe collect exceptions for telemetry?

**Comment by daxian-dbw** [comment]
> If it happens again, I'll run the script snippet you mentioned, and preserve the command history I was browsing through.

@cdmihai Yes, please do report back with those additional information if it happens to you again. Thanks!

@iSazonov I don't think the exception would be that interesting. A repro is what we desperately need :)

**Comment by iSazonov** [comment]
>  I don't think the exception would be that interesting. A repro is what we desperately need :)

My thoughts is that console applications is very sensitive and having all that we have with your script could be useful in telemetry.

**Comment by petacreepers23** [comment]
PS version: 5.1.18362.752
PSReadline version: 2.0.0-beta2
os: 10.0.18362.777 (WinBuild.160101.0800)
PS file version: 10.0.18362.777 (WinBuild.160101.0800)
HostName: ConsoleHost
BufferWidth: 100
BufferHeight: 500

Since I had same issue, this is the output of the script

**Comment by daxian-dbw** [comment]
@petacreepers23 thanks! You are using a fairly old version of PSReadLine, so the issue you saw should have been fixed by #1306 
Please upgrade to the 2.0.1 version of PSReadLine. If you run into the same after upgrading, please report back here with your environment data and repro steps.

**Comment by daniel0x00** [comment]
I see the same on PSReadLine 2.1.0-beta1:

### Environment
PSReadLine: 2.1.0-beta1
PowerShell: 7.0.0
OS: Microsoft Windows 10.0.17763
Last 200 Keys
```
 Ctrl+Backspace Backspace Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow Ctrl+LeftArrow $ redactedobject Spacebar = Spacebar Enter
 $ redactedobject Spacebar = Spacebar $ redactedobject . a r g s Enter
 $ redactedobject  Spacebar = Spacebar $ redactedobject  Spacebar | Spacebar s e l e c t Spacebar * , Spacebar @ { n = ' c o m m a n d ' ; e = { $ redactedobject  } } , Spacebar @ { n = ' REDACTED ' ; e = { $ e n v : C O M P U T E R N A M E } } Enter
 Enter
 $ redactedobject Enter
 Enter
 UpArrow UpArrow UpArrow DownArrow DownArrow UpArrow UpArrow UpArrow Escape
```

### Exception
```
System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')
   at System.Text.StringBuilder.Insert(Int32 index, String value)
   at Microsoft.PowerShell.PSConsoleReadLine.EditItemDelete.Undo()
   at Microsoft.PowerShell.PSConsoleReadLine.RevertLine(Nullable`1 key, Object arg)
   at Microsoft.PowerShell.PSConsoleReadLine.ProcessOneKey(PSKeyInfo key, Dictionary`2 dispatchTable, Boolean ignoreIfNoAction, Object arg)
   at Microsoft.PowerShell.PSConsoleReadLine.InputLoop()
   at Microsoft.PowerShell.PSConsoleReadLine.ReadLine(Runspace runspace, EngineIntrinsics engineIntrinsics, CancellationToken cancellationToken)

**Comment by jhoneill** [comment]
Another one here.  Identical stack trace displayed to @daniel0x00  above. 
It seems to be be fairly common and **easy to repro** in VS Code integrated shell

Put in a long line at the last line in the window, press [ESC] and POW! Something seems to do its sums wrongly and think it should be removing stuff from a line after the end of the window. 
If I move the cursor to left of about 176 (even if the line is very long) all is OK. If I'm not on the last line, all is OK. but cursor to near the bottom right is what triggers it .  Whilst very ugly it doesn't do any real _harm_  apart from displaying the error the session seems unaffected

```
PS version: 7.0.0
PSReadline version: 2.1.0-beta1'; ProjectUri = 'https://github.com/PowerShell/PSReadLine
os: 10.0.18362.1 (WinBuild.160101.0800)
PS file version: 7.0.0.0
HostName: Visual Studio Code Host
BufferWidth: 220
BufferHeight: 32
```

**Comment by daxian-dbw** [comment]
@jhoneill Can you please provide detailed repro steps? I tried with "a long line at the last line in the window, press [ESC]", but the exception didn't happen to me. So  a repro with exactly what you were typing would be really helpful. Thanks!

**Comment by jhoneill** [comment]
new window.  I'm multiple levels down so  prompt is 
`#500 PS7  C:\users\james\Documents\GitHub\vsteam\Source>` 
(so I'm already a fair distance to the right) 
I did dir 3 or 4 times to get to the bottom and fill the screen up then 
`
dir -recurse | select-string -pattern "lorem ipsum dolor amet" -SimpleMatch [esc]
`
Bang!

**Comment by iSazonov** [comment]
@jhoneill Could you please record video with repo steps?

**Comment by jhoneill** [comment]
@iSazonov  
Yesterday I could get this to happen at will on two different computers. And once it had happened I could restart the integrated shell and get it again.  Now it is more reluctant to happen. :-( Is there a bug which only happens after 12:00 noon ?  
I got one a few minutes ago and trying to show that it only happened if the cursor was on the last line I got this with the cursor well over to the right but not on the last line. The key that triggers it for me is always escape though. 

[vscode-crash.zip](https://github.com/PowerShell/PSReadLine/files/4558002/vscode-crash.zip)

**Comment by daxian-dbw** [comment]
Thanks @jhoneill. I found the root cause as well as a consistent repro, see https://github.com/PowerShell/PSReadLine/issues/1495#issuecomment-622471584
The issue will be fixed by #1496 (the 2nd commit).

That particular issue is different from what @cdmihai reported here initially, so further discussion should be done in #1495 instead.
